dbt_copilot_helper/utils/validation.py

```diff
@@ -142,3 +199,197 @@ def validator(string):
+)
+
+OPENSEARCH_ENGINE_VERSIONS = Or("2.5", "2.3", "1.3", "1.2", "1.1", "1.0")
+
+OPENSEARCH_INSTANCE_TYPES = Or(
+    "m6g.2xlarge.search",
+    "m6g.large.search",
+    "m6g.xlarge.search",
+    "t2.medium.search",
+    "t3.medium.search",
+)
+
+S3_BASE = {
+    Optional("readonly"): bool,
+    Optional("deletion-policy"): DELETION_POLICY,
+    Optional("services"): Or("__all__", [str]),
+    Optional("environments"): {
+        ENV_NAME: {
+            Optional("bucket-name"): Regex(
+                r"^(?!(^xn--|.+-s3alias$))^[a-z0-9][a-z0-9-]{1,61}[a-z0-9]$"
+            ),
+            Optional("deletion-policy"): Or("Delete", "Retain", "Snapshot"),
+        }
+    },
+}
+
+_S3_POLICY_DEFINITION = dict(S3_BASE)
+_S3_POLICY_DEFINITION.update({"type": "s3-policy"})
+S3_POLICY_SCHEMA = Schema(_S3_POLICY_DEFINITION)
+
+_S3_DEFINITION = dict(S3_BASE)
+_S3_DEFINITION.update(
+    {
+        "type": "s3",
+        Optional("objects"): [
+            {
+                "key": str,
+                Optional("body"): str,
+            }
+        ],
+    }
+)
+S3_SCHEMA = Schema(_S3_DEFINITION)
+
+AURORA_SCHEMA = Schema(
+    {
+        "type": "aurora-postgres",
+        "version": NUMBER,
+        Optional("deletion-policy"): DELETION_POLICY,
+        Optional("deletion-protection"): DELETION_PROTECTION,
+        Optional("environments"): {
+            ENV_NAME: {
+                Optional("min-capacity"): float_between_with_halfstep(0.5, 128),
+                Optional("max-capacity"): float_between_with_halfstep(0.5, 128),
+                Optional("snapshot-id"): str,
+                Optional("deletion-policy"): DELETION_POLICY,
+                Optional("deletion-protection"): DELETION_PROTECTION,
+            }
+        },
+        Optional("objects"): [
+            {
+                "key": str,
+                Optional("body"): str,
+            }
+        ],
+    }
+)
+
+RDS_SCHEMA = Schema(
+    {
+        "type": "rds-postgres",
+        "version": NUMBER,
+        Optional("deletion-policy"): DELETION_POLICY,
+        Optional("deletion-protection"): DELETION_PROTECTION,
+        Optional("environments"): {
+            ENV_NAME: {
+                Optional("plan"): RDS_PLANS,
+                Optional("instance"): RDS_INSTANCE_TYPES,
+                Optional("volume-size"): int_between(5, 10000),
+                Optional("replicas"): int_between(0, 5),
+                Optional("snapshot-id"): str,
+                Optional("deletion-policy"): DELETION_POLICY,
+                Optional("deletion-protection"): DELETION_PROTECTION,
+            }
+        },
+        Optional("objects"): [
+            {
+                "key": str,
+                Optional("body"): str,
+            }
+        ],
+    }
+)
+
+REDIS_SCHEMA = Schema(
+    {
+        "type": "redis",
+        Optional("deletion-policy"): DELETION_POLICY,
+        Optional("environments"): {
+            ENV_NAME: {
+                Optional("plan"): REDIS_PLANS,
```

Comment:
Probably outside the scope of this PR but some properties like "plan" and "engine" are required but not for every environment.

Maybe we need to enforce the "default" environment with the required properties?

Reply:
I think we should raise another ticket for this.